New column module in continuum mechanics

lfverlaan opened the pull request:
New Column module in continuum mechanics

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

"Fixes#28053"

#### Brief description of what is fixed or changed
A new module has been created named "column". This module aims to calculate reaction forces and displacements due to axial forces. The user can initialize a column with length, area and elastic modulus. Properties have been added so the user can adjust variables. Methods have been added where the user can apply supports and loads to the column. The module sets up the load equation using singularity functions.

#### Other comments
This is the first pull request for this module. The column module will be extended in future pull requests.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* physics.continuum_mechanics
   * Added a new `Column` class for modeling axially loaded structural elements, similar in structure to the `Beam` class. It supports applying axial point loads, fixed supports, and assembling the load equation using singularity functions.
<!-- END RELEASE NOTES -->

lfverlaan committed:
New files

Continuum mechanics: New files

Two new files have been added, column.py and a test file for the column script.

lfverlaan committed:
Update .mailmap

lfverlaan committed:
Update test_column.py

Tests have been written for the initialization of the column module, and the formulation of the load equation.

lfverlaan commented:
The tests have been committed. Now the functions will be written in the column.py file to make sure that the tests will run successfully.

Tom-van-Woudenberg commented on sympy/physics/continuum_mechanics/tests/test_column.py:
```diff
@@ -0,0 +1,79 @@
+    c1 = Column(20, E_1, A_1)
+
+    c_joined = c.join(c1)
+    assert c_joined.length == 40
```
Isn't this join thing a results of the old way of dealing with multiple parts? How is this altered by Macaulay's method in the beam implementation? @lfverlaan

lfverlaan replied:
I haven't looked much at how the beam module deals with different parts. I thought the function would be a good basis for if and when I would like to work with different areas and elastic module. I also felt that the join function could be a clear and user friendly manner to apply hinges with in the future. I guess that if this was an old method and there is a better way using macaulay's method I will have to research it further. Thanks!

Tom-van-Woudenberg commented:
> The tests have been committed. Now the functions will be written in the column.py file to make sure that the tests will run successfully.
> 
> The idea is to start with two boundary conditions (axial forces are zero just outside of the ends of the column) when initializing the class. When a support gets applied, a boundary condition for the deflection gets added. Now only simple instances will be explored, a column with two fixed ends and a column with one fixed and one free end. Any feedback on this approach?

Does the beam module function in a similar way (assuming zero axial forces)?

Would the functionality change when supports are added allongside the beam? If no other functions other required, it might be nice to add a test on that as well, making clear the scope of your implementation.

lfverlaan commented:
The beam module uses the conditions that the shear force and the moment just right of the end is equal to zero (with the limit functions). 

I would think that the functionality does not change if more supports are applied, only more deflection boundary conditions are then added. I am not sure that I understand your question perfectly, but do you mean that if there is no extra function needed for extra supports that these can directly tested? And that if columns with more than two supports can already be calculated that they should just directly be included in the scope?

Tom-van-Woudenberg commented:
> but do you mean that if there is no extra function needed for extra supports that these can directly tested?

Yes

> And that if columns with more than two supports can already be calculated that they should just directly be included in the scope?

Yes, or at least note that effectively your code works for a bit more than just the scope

lfverlaan committed:
Update test_column.py

Adjusted the tests. Added a test to show that adding three supports is also compatible. Also added a test to check if reaction forces due to supports are also added to the load equation. Also removed some tests, inputs that are outside of the scope.

lfverlaan committed:
Update column.py

The base of the Column class has been written. A column with a certain length, area and elastic modulus can be initialized. Methods have also been added to apply supports and forces. These methods add the singularity functions to the load equations. The user can get the load equation through the 'load' property.

lfverlaan commented:
A base for the Column class has been committed. The code has a structure very similar to the Beam module, as the necessary processes up to this point are largely the same. One major difference is that there is no distinction between types of supports; in the Column module, a support refers to any constraint that is fixed in the horizontal direction.

The main goals of this issue has been resolved: the load equation can now be set up for columns with multiple supports and point loads. An example use case of the module is shown below:

```
from sympy.physics.continuum_mechanics.column import Column
from sympy.core.symbol import (Symbol, symbols)

L, E, A, F = symbols('L E A F')
c = Column(L, E, A)

c.apply_support(0)
c.apply_support(L / 2)
c.apply_support(L)

c.apply_load(F, L / 4, -1)
c.apply_load(F, 3*L / 4, -1)

print(c.load)
```
> F*SingularityFunction(x, L/4, -1) + F*SingularityFunction(x, 3*L/4, -1) + R_0*SingularityFunction(x, 0, -1) + R_L*SingularityFunction(x, L, -1) + R_L/2*SingularityFunction(x, L/2, -1)

I would love hear any feedback!

Tom-van-Woudenberg commented:
Seems okay to me! Maybe add release note input here: https://github.com/sympy/sympy/pull/28049#issue-3045958183
@saiudayagiri might want to have a look too

If it's ready for a review by SymPy Core developers, please remove the draft label and tag Jason Moore

Tom-van-Woudenberg commented:
> ```
> from sympy.physics.continuum_mechanics.column import Column
> from sympy.core.symbol import (Symbol, symbols)
> 
> L, E, A, F = symbols('L E A F')
> c = Column(L, E, A)
> 
> c.apply_support(0)
> c.apply_support(L / 2)
> c.apply_support(L)
> 
> c.apply_load(F, L / 4, -1)
> c.apply_load(F, 3*L / 4, -1)
> 
> print(c.load)
> ```
> 
> > F_SingularityFunction(x, L/4, -1) + F_SingularityFunction(x, 3_L/4, -1) + R_0_SingularityFunction(x, 0, -1) + R_L_SingularityFunction(x, L, -1) + R_L/2_SingularityFunction(x, L/2, -1)
> 

Maybe you could add this example to your docstring?

Tom-van-Woudenberg committed:
Add docs

Tom-van-Woudenberg commented:
I proposed to add documentation here: https://github.com/lfverlaan/sympy/pull/1

lfverlaan committed:
Update column.py

A little bit of code cleanliness, removed whitespace

lfverlaan commented:
> I proposed to add documentation here: [lfverlaan#1](https://github.com/lfverlaan/sympy/pull/1)

Great! I will make sure to merge it into my branch

lfverlaan committed:
Merge pull request #1 from Tom-van-Woudenberg/docs

Add docs

lfverlaan committed:
Update column.py

Fixed a mistake where the reaction forces got added to the applied loads list. Now they are simply added to the load equation instead of the apply_load() function being called.

Also added examples in the docstring of the functions.

lfverlaan committed:
Update column.py

Checked the spelling, neatness.

lfverlaan commented:
The pull request is ready for review. 

@moorepants

Tom-van-Woudenberg committed:
Fix error in docs

Fixes Title overline too short

Tom-van-Woudenberg commented:
@lfverlaan , this fixes error in the docs: https://github.com/lfverlaan/sympy/pull/2

lfverlaan committed:
Merge pull request #2 from Tom-van-Woudenberg/docs

Fix error in docs

lfverlaan committed:
Update column.py

removed trailing whitespace

Tom-van-Woudenberg commented:
@saiudayagiri, since this is ready for review now, would you like to have a look at it?

saiudayagiri commented:
@Tom-van-Woudenberg, @lfverlaan, PR #28049 looks good so far! I’ve seen the join method in Beam but I’m not sure of its implementation details; I need to look into that.

Tom-van-Woudenberg commented:
> PR #28049 looks good so far! I’ve seen the join method in Beam but I’m not sure of its implementation details; I need to look into that.

@saiudayagiri, would you be able to review this now that you've been introduced properly?

saiudayagiri commented on sympy/physics/continuum_mechanics/column.py:
```diff
@@ -0,0 +1,275 @@
+
+        return reaction_load
+    
+    def apply_load(self, value, start, order):
```
I don’t see an end parameter in the apply_load method.which is necessary for distributed loads

saiudayagiri commented:
> > PR #28049 looks good so far! I’ve seen the join method in Beam but I’m not sure of its implementation details; I need to look into that.
> 
> @saiudayagiri, would you be able to review this now that you've been introduced properly?

Yes @Tom-van-Woudenberg and @lfverlaan — I’ve reviewed the PR #28049 and it looks good. From what I understand, this is primarily being developed for Structure2D integration, so the implementation seems appropriate. I did notice that apply_load currently lacks an end parameter, but I see that’s addressed in the next PR. The removal of the .join() method also makes sense, as it isn't necessary.

saiudayagiri commented on sympy/physics/continuum_mechanics/column.py:
```diff
@@ -0,0 +1,275 @@
+
+        length: Sympifyable
+            A Symbol or value representing thecColumn's length.
+        
```
Solve this issue as this got reviewed and open

Tom-van-Woudenberg commented:
Superseded by https://github.com/sympy/sympy/pull/28138. @lfverlaan , could you close this one?
